validator/sawtooth_validator/state/merkle.py

```diff
@@ -307,11 +166,114 @@ def addresses(self):
 
         return addresses
 
-    def leaves(self, prefix):
-        leaves = {}
-        for address, value in self._yield_iter(prefix, self._root_hash):
-            leaves[address] = value
-        return leaves
+    def leaves(self, prefix=None):
+        """Returns an iterator which returns tuples of (address, data) values
+        """
+        try:
+            return _LeafIterator(self._merkle_db_ptr, prefix)
+        except KeyError:
+            # The prefix doesn't exist
+            return iter([])
 
     def close(self):
-        self._database.close()
+        pass
+
+
+def _prepare_byte_result():
+    return (ctypes.POINTER(ctypes.c_uint8)(), ctypes.c_size_t(0))
+
+
+def _from_c_bytes(c_data, c_data_len):
+    # pylint: disable=invalid-slice-index
+    return bytes(c_data[:c_data_len.value])
+
+
+def _libexec(name, *args):
+    res = LIBRARY.call(name, *args)
+    if res == ErrorCode.Success:
+        return
+    elif res == ErrorCode.NullPointerProvided:
+        raise TypeError("Provided null pointer(s)")
+    elif res == ErrorCode.NotFound:
+        raise KeyError("Value was not found")
+    elif res == ErrorCode.DatabaseError:
+        raise ValueError("A Database Error occurred")
+    elif res == ErrorCode.InvalidHashString:
+        raise KeyError(
+            "merkle root was not a valid hash")
+    elif res == ErrorCode.InvalidAddress:
+        raise KeyError(
+            "Address was not valid ")
+    elif res == ErrorCode.StopIteration:
+        raise StopIteration()
+    elif res == ErrorCode.Unknown:
+        raise ValueError("An unknown error occurred")
+    else:
+        raise ValueError("An unknown error occurred: {}".format(res))
```

Comment:
This error code to exception mapping seems like it would be common enough to warrant providing a function in sawtooth_validator.ffi to simplify it. It might also be a good idea to define a base error type that includes some common exceptions like `Success` and `NullPointer`.

Reply:
I didn't combine these into a common one because there's no way to do the same on the Rust side of things.  The error codes are just enums.